client/extensions/woocommerce/app/product-categories/index.js

```diff
@@ -1,29 +1,113 @@
 /**
  * External dependencies
  */
-
 import React, { Component } from 'react';
+import { bindActionCreators } from 'redux';
 import classNames from 'classnames';
 import { connect } from 'react-redux';
 import { localize } from 'i18n-calypso';
+import { union, includes, trim, debounce } from 'lodash';
 
 /**
  * Internal dependencies
  */
 import ActionHeader from 'woocommerce/components/action-header';
+import { DEFAULT_PRODUCT_CATEGORIES_PER_PAGE } from 'woocommerce/state/sites/product-categories/utils';
+import { fetchProductCategories } from 'woocommerce/state/sites/product-categories/actions';
 import { getLink } from 'woocommerce/lib/nav-utils';
 import { getSelectedSiteWithFallback } from 'woocommerce/state/sites/selectors';
 import Main from 'components/main';
 import NavTabs from 'components/section-nav/tabs';
 import NavItem from 'components/section-nav/item';
+import ProductCategoriesList from 'woocommerce/app/product-categories/list';
 import SectionNav from 'components/section-nav';
 import Search from 'components/search';
 
 class ProductCategories extends Component {
 
+	state = {
+		requestedPages: [ 1 ],
+		requestedSearchPages: [],
+	};
+
+	constructor( props ) {
+		super( props );
+		this.debouncedOnSearch = debounce( this.onSearch, 500 );
+	}
+
+	componentWillMount() {
+		const { site } = this.props;
+		this.props.fetchProductCategories( site.ID, {
+			page: 1,
+			per_page: DEFAULT_PRODUCT_CATEGORIES_PER_PAGE,
+		} );
+	}
+
+	componentWillReceiveProps( newProps ) {
+		const { site } = this.props;
+		const newSiteId = ( newProps.site && newProps.site.ID ) || null;
+		const oldSiteId = ( site && site.ID ) || null;
+		if ( oldSiteId !== newSiteId ) {
+			this.props.fetchProductCategories( newSiteId, {
+				page: 1,
+				per_page: DEFAULT_PRODUCT_CATEGORIES_PER_PAGE,
+			} );
+		}
+	}
+
+	requestPages = pages => {
+		const { site } = this.props;
+		const { searchQuery } = this.state;
+
+		if ( searchQuery && searchQuery.length ) {
+			pages.forEach( page => {
+				if ( ! includes( this.state.requestedSearchPages, page ) ) {
```

Comment:
Since there is a bit of duplicated logic in both parts of this `if/else` block, should we just create a `query` object at the top level, and add a `search: searchQuery` attribute to it if there is a searchQuery? Then the looping of pages and setState call could just be combined.